tests/test_edge_cases.py

```diff
@@ -1340,6 +1342,8 @@ class TopModel(BaseModel):
         self: str
         nest: NestedModel
 
+    # TODO: Do we want any changes to this behavior in v2? (Currently the __init__-override is not called)
+    #   We could detect if `__init__` has been overridden if necessary...
     m = TopModel.model_validate(dict(self='Top Model', nest=dict(self='Nested Model', modified_number=0)))
```

Comment:
Question 3 ☝️

Reply:
I guess this should be an error or warning. If you want to do stuff on init, you should use `model_post_init`

Reply:
What if you want to do something in `__init__` _prior_ to the call to `super().__init__()`?

An example might be adding an extra argument that is used to do something to the other input arguments (e.g., something like having a flag to indicate whether they should be manually injected from environment if not present).

I'm not saying this is necessarily something we should intend to support, but I could see it posing a migration challenge for users who are using this functionality with their existing (v1) BaseModel classes.

If there's a way to migrate such logic I think it's fine to require changes, just seems like a good idea to make sure it's possible. (Seems like something a root-validator would be able to do, but not sure what the status of those is in v2).

Reply:
I'm curious as to why this doesn't work, I assume it's a technical reason?

Reply:
> why this doesn't work

from `pydantic.main`:
```python
    @classmethod
    def model_validate(cls: type[Model], obj: Any) -> Model:
        values, fields_set = cls.__pydantic_validator__.validate_python(obj)
        m = cls.__new__(cls)
        _object_setattr(m, '__dict__', values)
        _object_setattr(m, '__fields_set__', fields_set)
        if hasattr(cls, '__pydantic_post_init__'):
            cls.__pydantic_post_init__(context=None)  # type: ignore[attr-defined]
        return m
```
You can see it doesn't actually call the `__init__` function.